subsys/nfc/t4t/isodep.c

```diff
@@ -0,0 +1,879 @@
+		isodep_r_frame_send(true);
+
+		return;
+
+	case ISODEP_FRAME_WTX_RESPONSE:
+		if (err_status->frame_retry_cnt > CONFIG_NFC_T4T_ISODEP_WTX_RETRY) {
+			break;
+		}
+
+		/* Resend last S(WTX) response. Last Response should be in
+		 * Tx buffer.
+		 */
+		if (t4t_isodep_cb->data_send) {
+			t4t_isodep_cb->data_send(t4t_isodep.tx_data.data,
+						 t4t_isodep.tx_data.len,
+						 err_status->wtx * t4t_isodep.tag.fwt +
+						 T4T_FWT_DELTA + NFCA_T4T_FWT_T_FC);
+		}
+
+		return;
+
+	case ISODEP_FRAME_DESELECT:
+		if (err_status->frame_retry_cnt > CONFIG_NFC_T4T_ISODEP_DESELECT_RETRY) {
+			break;
+		}
+
+		/* Resend last S(DESELECT) command. Last command should be in
+		 * Tx buffer.
+		 */
+		if (t4t_isodep_cb->data_send) {
+			t4t_isodep_cb->data_send(t4t_isodep.tx_data.data,
+						 t4t_isodep.tx_data.len,
+						 ISODEP_FWT_DEACTIVATION);
+		}
+
+		return;
+
+	default:
+		break;
+	}
+
+	err = timeout ? -NFC_T4T_ISODEP_TIMEOUT_ERROR : -NFC_T4T_ISODEP_TRANSMISSION_ERROR;
+
+	if (err) {
+		atomic_set(&t4t_isodep.state, ISODEP_STATE_INITIALIZED);
+		err_notify(err);
+	}
+}
+
+static void isodep_delay_handler(struct k_work *work)
+{
+	isodep_chunk_send();
+}
+
+int nfc_t4t_isodep_init(u8_t *tx_buf, size_t size_tx,
+			u8_t *rx_buf, size_t size_rx,
+			const struct nfc_t4t_isodep_cb *cb)
+{
+	if (!tx_buf || !rx_buf) {
+		return -EINVAL;
+	}
+
+	if (size_tx < T4T_FSD_MIN) {
+		LOG_ERR("Tx buffer is to small. Minimum Tx buffer size %d\n",
+			T4T_FSD_MIN);
+		return -EINVAL;
+	}
+
+	if (size_rx < T4T_FSD_MIN) {
+		LOG_ERR("Rx buffer is to small. Minimum Rx buffer size %d\n",
+			T4T_FSD_MIN);
+		return -EINVAL;
+	}
+
+	if (atomic_get(&t4t_isodep.state) != ISODEP_STATE_UNINITIALIZED) {
+		LOG_ERR("Already initialized");
+
+		return -EACCES;
+	}
+
+	t4t_isodep.tx_data.data = tx_buf;
+	t4t_isodep.tx_data.buf_size = size_tx;
+	t4t_isodep.rx_data.data = rx_buf;
+	t4t_isodep.rx_data.buf_size = size_rx;
+
+	t4t_isodep_cb = cb;
+
+	/** Initialize work queue to handle delay before sending first
+	 * after receipt RATS Response.
+	 */
+	k_delayed_work_init(&isodep_work, isodep_delay_handler);
+
+	atomic_set(&t4t_isodep.state, ISODEP_STATE_INITIALIZED);
+
+	return 0;
+}
+
+void nfc_t4t_isodep_on_timeout(void)
+{
+	isodep_error_handle(true);
+}
+
+int nfc_t4t_isodep_rats_send(enum nfc_t4t_isodep_fsd fsd, u8_t did)
+{
+	u8_t param;
+	u32_t state = atomic_get(&t4t_isodep.state);
+
+	if ((state == ISODEP_STATE_UNINITIALIZED) ||
+	    (state == ISODEP_STATE_TRANSFER)) {
+		return -EACCES;
+	}
+
+	atomic_set(&t4t_isodep.state, ISODEP_STATE_TRANSFER);
+
+	if (did > T4T_DID_MAX) {
+		LOG_ERR("Invalid DID value. It should be between 0-14.");
+
+		return -EINVAL;
+	}
+
+	if (t4t_isodep.tx_data.buf_size < fsd_value_map[fsd]) {
+		LOG_ERR("Invalid FSD value. Increase Tx buffer size or decrease FSD");
+
+		return -ENOMEM;
+	}
+
+	/* Set DID field. */
+	param = did & T4T_RATS_DID_MASK;
+
+	/* Set FSDI field. */
+	param = (fsd << T4T_RATS_FSDI) & T4T_RATS_FSDI_MASK;
```

Comment:
Should be `|=` otherwise did will be cleared

Reply:
Yes right. Fixed